Add an alternative parquet reader implementation

michoecho opened the pull request:
As discussed, I've implemented an alternative parquet reader library, and a Parquet to CQL converter on top of it.

I've briefly described all the files and their order in README.md.

The changes are not production ready. In particular, some compression algorithms are not implemented yet (that one's easy), there are no comprehensive tests, and most importantly, I'm not confident in the input validation, because right now, I validate the metadata ad-hoc - that is, I check whether a piece of metadata makes sense only when I'm about to use it. It's too easy to miss some crucial check (e.g. size >= 0) that way, and I don't trust that I haven't done just that somewhere. Nevertheless, I think the code is ready for review.

Commit:
Add an alternative parquet reader implementation

psarna commented on src/parquet/file_reader.cc:
```diff
@@ -0,0 +1,764 @@
+        DeserializedType& deserialized_msg
+) {
+    using ThriftBuffer = apache::thrift::transport::TMemoryBuffer;
+    auto tmem_transport = std::make_shared<ThriftBuffer>(const_cast<uint8_t*>(serialized_msg), *serialized_len);
```
This line is a cause of failure for me when I compile with Thrift 0.14, because it tries to free `serialized_msg` in the destructor... no idea why. It works when I artificially allocate memory here and copy, but that's not how it *should* work at all, since the default constructor of TMemoryBuffer does not take ownership of the pointer.

That's quite frustrating, since the stock Thrift for Fedora is 0.10 and it requires using `boost::shared_ptr` instead of `std::shared_ptr`, which makes `parquet/` compilation impossible, and Thrift versions 0.11-0.13 produce compilation errors themselves, so I can't easily build and install them... Btw, we need to add a CMake line that ensures Thrift is at least 0.11, otherwise `parquet/` does not compile due to not being able to assign std::shared_ptr to boost::shared_ptr.

I'll debug that tomorrow, or try to find a Thrift version that works (unless it's somehow an error in this patch, but I assume you'd notice that the test always fails with a segfault).

psarna reviewed (commented):
Thanks, I only made a cursory review, I'll continue on Sunday - I left one thrift-related commit so far.

General things to do later, before it can go upstream:
1. Break it into a few separate commits (e.g. split testing, CMake preparation for Thrift, etc. to separate commits)
2. Provide commit messages for each (https://github.com/scylladb/scylla/wiki/Formatting-and-sending-patches)

psarna commented:
update: my fault, custom thrift still used stock headers, which caused it to use wrong offsets for member attributes.

As for yesterday's comment - one more thing to split from this big patch would be the cql_reader - ideally, file_reader should exist in its own patch, and the cql_reader should be an optional app. But, again, that's for later, the patch in its current state is good for review as well.

psarna commented on cmake/FindThrift.cmake:
```diff
@@ -0,0 +1,102 @@
+  set_target_properties(Thrift::thrift
+                        PROPERTIES IMPORTED_LOCATION "${THRIFT_STATIC_LIB}"
+                                   INTERFACE_INCLUDE_DIRECTORIES "${THRIFT_INCLUDE_DIR}")
+  extract_thrift_version()
```
I think we should consider a check that thrift is >= 11 here - otherwise thrift may be based on `boost::shared_ptr`, which does not match the rest of this patch.

michoecho replied:
Fixed in 6cbc79e.

psarna commented on include/seastar/parquet/bpacking.hh:
```diff
@@ -0,0 +1,3819 @@
+#pragma once
```
I wonder if it passes our review :) It kind of looks like something that compiler would optimize anyway. But since it's an external, machine-generated file, let it stay for now. It should probably have a copy of its original license somewhere on top.

michoecho replied:
Added a licence notice in e3555d1.

psarna commented on include/seastar/parquet/exception.hh:
```diff
@@ -0,0 +1,43 @@
+        if (!msg.empty()) {
+            ss << ": " << msg << ".";
+        }
+        return parquet_exception(ss.str());
```
nitpick - it's an exception, so it's considered to be rare, so it's fine to use a helper format function, which makes the code more readable:
```cpp
return parquet_exception(format("Unexpected end of stream: {}", msg));
```
Also, let's not use default arguments - that applies to all exceptions below as well. Using a default empty msg makes debugging hard, and lack of a default argument will force engineers to put something meaningful in the message.

michoecho replied:
Changed in e23d38a.

psarna commented on include/seastar/parquet/exception.hh:
```diff
@@ -0,0 +1,43 @@
+        return parquet_exception(ss.str());
+    }
+
+    static parquet_exception nyi(const std::string& msg = "") {
```
let's expand this name to `not_yet_implemented`, `nyi` is not a well-known acronym. Also - remove the default, as above and feel free to use seastar's `format()` instead of a raw string stream

michoecho replied:
Thanks for pointing me to seastar::format. A format function is a blessing for writing error messages. I used it extensively in e23d38a.

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,475 @@
+
+namespace parquet {
+
+////////////////////////////////////////////////////////////////////////////////
```
nitpick: these separating dashes will probably need to go, as they're not part of our codestyle

michoecho replied:
Fixed in ac9cdae.

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,475 @@
+    std::unique_ptr<uint8_t[]> _data;
+    static constexpr inline uint64_t next_power_of_2(uint64_t n) {
+        if (n < 2) return n;
+        return 1ull << (64 - __builtin_clzll(n - 1));
```
Take a look at `include/seastar/core/bitops.hh` - we have wrappers for clz (`count_leading_zeros` and also `log2ceil` and `log2floor`, which is precisely what you can apply here), and it's preferable to use them, in case we want to support different architectures in the future.

michoecho replied:
Fixed in f38fe11.
> it's preferable to use them, in case we want to support different architectures in the future

Yeah, sure...

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,475 @@
+    int _values_read;
+    int bit_width(int max_n) {
+        assert (max_n >= 0);
+        return (max_n == 0) ? 0 : (32 - __builtin_clz(static_cast<int32_t>(max_n)));
```
ditto: `include/seastar/core/bitops.hh`

michoecho replied:
Fixed in f38fe11.

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,475 @@
+
+    // Read a batch of n levels (the last batch may be smaller than n).
+    template <typename T>
+    size_t read_batch(int n, T *out) {
```
nitpick: in our code style, `*` sticks to the type: `T* out`

michoecho replied:
Fixed in ac9cdae.

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,475 @@
+////////////////////////////////////////////////////////////////////////////////
+
+/* Refer to the parquet documentation for the description of various encodings:
+ * https://github.com/apache/parquet-format/blob/master/Encodings.md
```
If the documentation is also apache-licensed, and I assume it is, it's probably a good idea to copy this file to our source code in `doc/parquet/`, just in case it disappears one day.

michoecho replied:
Done in 12d4c91.

psarna commented on include/seastar/parquet/overloaded.hh:
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Feel free to put this one in `include/seastar/util`, wrap it in a separate patch and send it to sestar-dev - we do have `overloaded` in scylla already, but seastar is eventually going to need it too.

psarna commented on include/seastar/parquet/parquet_types.h:
```diff
@@ -0,0 +1,2915 @@
+/**
+ * Autogenerated by Thrift Compiler (0.13.0)
```
Ok - so this one should either be generated during building or we need to set a version check in CMake, as mentioned above in my comment for FindThrift.

michoecho replied:
Fixed in 6cbc79e.

psarna commented on src/parquet/file_reader.cc:
```diff
@@ -0,0 +1,764 @@
+ * Our strategy (rewind only when _buffer_start moves past half of buffer.size()) guarantees that
+ * we will actively use at least 1/2 of allocated memory, and that any given byte is rewound at most once.
+ */
+void peekable_stream::ensure_space(size_t n) {
```
We talked about unit tests before - that function is one thing that could be unit-tested to ensure that we allocate appropriate amounts of memory for corner cases, just like the one you explained in the comment above.

psarna commented on src/parquet/file_reader.cc:
```diff
@@ -0,0 +1,764 @@
+        if (_buffer.size() == 0) {
+            return i;
+        }
+        if (4 > _buffer.size()) {
```
that's a Yoda condition - let's use human-readable `_buffer.size() < 4` instead

michoecho replied:
Fixed in ac9cdae.

psarna commented on src/parquet/file_reader.cc:
```diff
@@ -0,0 +1,764 @@
+
+        // TODO: the key to this algorithm is to decode the entire miniblock at once.
+        int64_t delta;
+        if (!_decoder.GetValue(_delta_bit_width, &delta)) throw parquet_exception::eof();
```
please wrap all `if` statements in curly braces, that's our strict code style

michoecho replied:
Fixed in ac9cdae.

psarna commented on src/parquet/file_reader.cc:
```diff
@@ -0,0 +1,764 @@
+ * the members that have been defined at this point.
+ * https://en.cppreference.com/w/cpp/language/class_template
+ *
+ * Has value_decoder<T> not been "defined at this point" or what?
```
Can't help here, but it's quite possible you'll get your answers once these patches are sent to seastar-dev.

psarna reviewed (commented):
Impressive, thanks. Left some comments.

Commit:
Fix a minor logic error in cql_reader

Commit:
Add thrift version check

Commit:
Fix code style issues in file_reader

Commit:
Copy license notices to files borrowed from Arrow

Commit:
Revise error handling

Commit:
Use seastar wrappers for __builtin_clz

michoecho commented:
I have addressed most of the comments.
Besides that, I have revised my error handling in e23d38a - I made the error messages more descriptive, added some outer try...catch wrappers to add context to the errors (i.e. in which column invalid data was found), added some missing validity checks, and changed some integer types. (Sorry for making another big, non-atomic commit, I will be more disciplined about that next time).

Commit:
Copy chosen parquet docs in

Commit:
Add documentation references

psarna commented on cmake/SeastarDependencies.cmake:
```diff
@@ -102,6 +109,7 @@ macro (seastar_find_dependencies)
   set (_seastar_dep_args_lksctp-tools REQUIRED)
   set (_seastar_dep_args_rt REQUIRED)
   set (_seastar_dep_args_yaml-cpp 0.5.1 REQUIRED)
+  set (_seastar_dep_args_Thrift 0.11.0 REQUIRED)
```
Unless I'm mistaken, that line **unconditionally** sets thrift as required, while thirft itself is only present if `Seastar_EXPERIMENTAL_PARQUET` is enabled - that would need to be changed, so that we only require thrift version if we require thrift in the first place.

michoecho replied:
You happen to be mistaken - if `Seastar_EXPERIMENTAL_PARQUET` is disabled, Thrift is not included in the _seastar_all_dependencies list. Hence `find_package` is never called for Thrift at all, (and the `_seastar_dep_args_*` variables are only there to be passed to `find_package` as a version check).

psarna replied:
:ok_hand: I stand corrected

psarna commented on include/seastar/parquet/file_reader.hh:
```diff
@@ -0,0 +1,482 @@
+        return seastar::make_ready_future<size_t>(0);
+    }
+    if (!_initialized) {
+        return load_next_page().then([=] {
```
This comment may apply to other places as well: Seastar does not accept default capturing, except capturing by reference for obvious cases (e.g. in a lambda passed to `std::all_of`) - continuation chains are, however, much more sensitive to object liveness, so we require an explicit capture list, e.g. `[n, def, rep, val]`.

michoecho replied:
Fixed in #92.

psarna commented on cmake/SeastarDependencies.cmake:
```diff
@@ -75,6 +75,13 @@ macro (seastar_find_dependencies)
     rt
     yaml-cpp)
 
+  if (Seastar_EXPERIMENTAL_PARQUET)
```
With thrift, we need to change the order of commits - this one needs to go first. That's because we strive to keep git history bisectability - i.e. each time you check out a seastar commit from a repo, it should compile and pass all tests. If thrift is checked only after we introduce thrift-related changes, we have a hole in git history in which we try to compile thrift-dependent code without ensuring its version first.

michoecho replied:
Fixed in #92.

psarna reviewed (commented):
Thanks. Looks better, but I still left some comments and I have a more generic comment about this whole series. The changes I'm going to request are not related to code, so it may be a little frustrating, but that's how big projects work :) We need to adhere to certain ways in order to keep the code maintainable. So:
1. This series of patches introduces a single, self-contained functionality - that's good.
2. For a series to be mergeable, we require that it does not contain any fixes to itself - e.g. a first commit introduces a big chunk of code, and then commits 2, 4, 6 and 7 fix some minor issues with 1 - that's what we try to avoid, because later when we need to backport or bisect for an error, that makes debugging way easier. It also makes the series easier to review.
3. It's good that the following were moved to separate patches:
   * 6cbc79e2d3a7513ca9476e55449db2a7e4ebe529
   * e3555d1af301d357d2fd2e563a7ac573b4747fb7
   * 12d4c910c8d46cea63c96bc153470833f970e8c1
   * fae23b22d9c0ee8f42827aef4d942bc856b65bf4
... however, one more split is in order, the most important one - there should be a separate commit that introduces `file_reader` and its helpers (`record_reader`, etc), because that's a self-contained functionality which allows reading Parquet files. Then, a separate commit should introduce `cql_reader`, which is a utility built on top of `file_reader`, in order to generate CQL queries from Parquet. It's very likely that the whole `cql_reader` will later be moved to `/apps` directory, and having it in a separate commit makes it easier. Still, the most important reason is that it's a separate entity - `file_reader` does not need any `cql_reader` capabilities. While we're at splitting anyway, I think it would be good to also have all the C++ code copied from arrow as a separate commit, which prepares needed files for the `file_reader` commit. So, the structure would look like:
1. A commit which adds stuff copied from arrow
2. A commit which adds file_reader
3. A commit which adds file_reader tests (can be squashed with 2)
4. A commit which adds cql_reader

Again, I know it sounds like moving code around without much reason, but it greatly helps the maintainability of Seastar to enforce strict rules - and in order for these patches to land upstream, we need to follow these rules.

That said, the code itself looks very good, so once we get these patches in order and split properly, we're going to send them upstream soon. Thanks!

prawilny committed:
Fix 8-bit integer printing into CQL

prawilny committed:
Enable already implemented gzip decompression in file reader

prawilny committed:
Add CQL reader test: all types with different compressions
